lib/opam_solve.ml

```diff
@@ -187,37 +221,87 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
 
   module Solver = Opam_0install.Solver.Make (Context)
 
-  let build_context ~build_only ~allow_jbuilder ~ocaml_version ~local_packages
-      ~pin_depends ~target_packages input =
+  let build_context ~build_only ~allow_jbuilder ?require_dune ?exempt_dune_for
+      ~ocaml_version ~local_packages ~pin_depends ~target_packages input =
     let open Result.O in
     let install_test_deps_for =
       if build_only then OpamPackage.Name.Set.empty else target_packages
     in
     let constraints = constraints ~ocaml_version in
     let+ fixed_packages = fixed_packages ~local_packages ~pin_depends in
-    Context.create ~install_test_deps_for ~allow_jbuilder ~constraints
-      ~fixed_packages input
+    Context.create ~install_test_deps_for ~allow_jbuilder ?require_dune
+      ?exempt_dune_for ~constraints ~fixed_packages input
+
+  let opam_provided opam_chunk =
+    let open Result.O in
+    let* names =
+      match opam_chunk.OpamParserTypes.FullPos.pelem with
+      | OpamParserTypes.FullPos.String s -> Ok [ s ]
+      | List items ->
+          items.OpamParserTypes.FullPos.pelem
+          |> List.map ~f:(fun item ->
+                 match item.OpamParserTypes.FullPos.pelem with
+                 | OpamParserTypes.FullPos.String s -> Ok s
+                 | _ -> Error "String required")
+          |> Result.List.all
+      | _ -> Error "String or List required"
+    in
+    let names = List.map ~f:OpamPackage.Name.of_string names in
+    Ok (OpamPackage.Name.Set.of_list names)
+
+  let determine_dependencies context request =
+    match Solver.solve context request with
+    | Error e -> Error (`Diagnostics e)
+    | Ok selections ->
+        Solver.packages_of_result selections
+        |> List.map ~f:(fun pkg -> pkg.OpamPackage.name)
+        |> OpamPackage.Name.Set.of_list |> Result.ok
+
+  type raw_calculation = { package : OpamPackage.t; vendored : bool }
 
-  let calculate_raw ~local_packages ~target_packages context =
-    let target_packages_names = OpamPackage.Name.Set.elements target_packages in
+  let pp_raw_calculation fmt { package; vendored = _ } =
+    Opam.Pp.package fmt package
+
+  let opam_provided_packages local_packages target_packages =
```

Comment:
I don't think this belongs in the solver, this should rather be extracted from `Lock`. It should also use the `Source_opam_config` I think, it would simplify things a bit here.

Reply:
I thought it wouldn't but after moving stuff to use `Source_opam_config` you're right - this makes more sense in `Lock`. Only downside is another argument to the `Solver.compute` but I guess from a logical perspective it should not be part of the computation of dependencies but rather injected into it, so this is indeed better!